src/lib/Monads.ml

```diff
@@ -131,48 +132,34 @@ struct
   let abort_if_inconsistent : 'a -> 'a m -> 'a m =
     fun abort m ->
     fun st ->
-    match CofEnv.status st.cof_env with
+    match CofEnv.already_inconsistent st.cof_env with
     | `Consistent -> m st
     | `Inconsistent -> M.ret abort st
 
   let lift_cmp (m : 'a compute) : 'a m =
     fun {state; veil; cof_env} ->
     m {state; veil; cof_env}
 
+  let replace_env cof_env m =
+    M.scope (fun local -> {local with cof_env}) @@
+    abort_if_inconsistent `Abort @@
+    let+ x = m in
+    `Ret x
+
   let restrict phi m =
     let* {cof_env} = M.read in
-    let cof_env = CofEnv.assume cof_env phi in
-    match CofEnv.status cof_env with
-    | `Consistent ->
-      M.scope (fun local -> {local with cof_env}) @@
-      let+ x = m in
-      `Ret x
-    | `Inconsistent ->
-      M.ret `Abort
-
-  let left_invert_under_cof phi m =
-    let rec go cofs m =
-      abort_if_inconsistent () @@
-      match cofs with
-      | [] -> m
-      | (Cof.Var _ | Cof.Cof (Cof.Eq _)) as phi :: cofs ->
-        begin
-          restrict phi @@ go cofs m |>> fun _ -> M.ret ()
-        end
-      | Cof.Cof (Cof.Meet phis) :: cofs ->
-        go (phis @ cofs) m
-      | Cof.Cof (Cof.Join phis) :: cofs ->
-        MU.app (fun phi -> go (phi :: cofs) m) phis
-    in
-    go [phi] m
-
-  let left_invert_under_current_cof m =
-    let* env = M.read in
-    left_invert_under_cof (CofEnv.unreduced_assumptions env.cof_env) m
+    replace_env (CofEnv.assume cof_env phi) m
+
+  let left_invert_under_cofs phis m =
+    let* {cof_env} = M.read in
+    CM.left_invert_under_cofs cof_env phis @@
+    fun cof_env ->
+    let+ _ = replace_env cof_env m in ()
```

Comment:
is the let+ doing anything here? or can we eta reduce this?

Reply:
`garbage |>> fun _ -> M.ret ()`

We can put this into `Monad.Util`, I guess. The monadic version of `ignore`.

Reply:
oh i see.  i think the code is ok then